qiskit/backends/_sympy_unitarysimulator.py

```diff
@@ -0,0 +1,280 @@
+            self._configuration = {'name': 'local_sympy_unitary_simulator',
+                                   'url': 'https://github.com/QISKit/qiskit-sdk-py',
+                                   'simulator': True,
+                                   'local': True,
+                                   'description': 'A python simulator for unitary matrix',
+                                   'coupling_map': 'all-to-all',
+                                   'basis_gates': 'u1,u2,u3,cx,id'}
+        else:
+            self._configuration = configuration
+        self._unitary_state = None
+        self._number_of_qubits = None
+
+    # the following is the sympy implementations of various gates, including basic gates and ugates.
+    @staticmethod
+    def regulate(theta):
+        """Find the most close symbolic form"""
+        error_margin = 0.01
+
+        if abs(N(theta - pi)) < error_margin:
+            return pi, True
+        if abs(N(theta - pi/2)) < error_margin:
+            return pi/2, True
+        if abs(N(theta - pi/4)) < error_margin:
+            return pi/4, True
+        if abs(N(theta - 2*pi)) < error_margin:
+            return 2*pi, True
+
+        return theta, theta == 0  # if theta ==0, we also think it is regular
+
+    @staticmethod
+    def compute_ugate_matrix(parafloatlist):
+        """compute the ugate matrix
+            Args:
+                parafloatlist(list): parafloatlist
+            Returns:
+                Matrix: the matrix that represents the U gate
+        """
+        theta = parafloatlist[0]
+        phi = parafloatlist[1]
+        lamb = parafloatlist[2]
+
+        theta, theta_is_regular = SympyUnitarySimulator.regulate(theta)
+        phi, phi_is_regular = SympyUnitarySimulator.regulate(phi)
+        lamb, lamb_is_regular = SympyUnitarySimulator.regulate(lamb)
+        left_up = cos(theta/2)
+        right_up = (-E**(I*lamb)) * sin(theta/2)
+        left_down = (E**(I*phi)) * sin(theta/2)
+        right_down = (E**(I*(phi+lamb))) * cos(theta/2)
+        u_mat = Matrix([[left_up, right_up], [left_down, right_down]])
+
+        if theta_is_regular and phi_is_regular and lamb_is_regular:
+            u_mat_numeric = u_mat
+        else:
+            u_mat_numeric = u_mat.evalf()
+        return u_mat_numeric
+
+    @staticmethod
+    def compute_ugate_matrix_wrap(parafloatlist):
+        """preparation needed before computing the ugate matrix
+            Args:
+                parafloatlist (list): parafloatlist
+            Returns:
+                Matrix: matrix that represents the ugate
+        """
+        if len(parafloatlist) == 1:  # [theta=0, phi=0, lambda]
+            parafloatlist.insert(0, 0.0)
+            parafloatlist.insert(0, 0.0)
+        elif len(parafloatlist) == 2:  # [theta=pi/2, phi, lambda]
+            parafloatlist.insert(0, pi/2)
+        elif len(parafloatlist) == 3:  # [theta, phi, lambda]
+            pass
+        else:
+            return NotImplemented
```

Comment:
maybe raise an alarm/exception here?

Reply:
you mean in L253?

Reply:
done. U gates must carry 1,2, or 3 parameters. Otherwise, we do the following: 
                raise Exception('U gate must carry 1, 2 or 3 parameters!')